source/dofs/dof_handler_policy.cc

```diff
@@ -5591,6 +5591,103 @@ namespace internal
+                  partitioner.n_ghost_indices()),
+                requests);
+              partitioner.export_to_ghosted_array_finish(
+                ArrayView<types::global_dof_index>(
+                  ghosted_new_numbers.data() + new_numbers.size(),
+                  partitioner.n_ghost_indices()),
+                requests);
+
+              // we need to fill the indices of the locally owned part into
+              // the new numbers array. their right position is somewhere in
+              // the middle of the array, so we first copy the ghosted part
+              // from smaller ranks to the front, then insert the data in the
+              // middle.
+              unsigned int n_ghosts_on_smaller_ranks = 0;
+              for (std::pair<unsigned int, unsigned int> t :
+                   partitioner.ghost_targets())
+                {
+                  if (t.first > partitioner.this_mpi_process())
+                    break;
+                  n_ghosts_on_smaller_ranks += t.second;
+                }
+              if (n_ghosts_on_smaller_ranks > 0)
+                {
+                  Assert(ghosted_new_numbers.data() != nullptr,
+                         ExcInternalError());
+                  std::memmove(ghosted_new_numbers.data(),
+                               ghosted_new_numbers.data() + new_numbers.size(),
+                               sizeof(types::global_dof_index) *
+                                 n_ghosts_on_smaller_ranks);
+                }
+              if (new_numbers.size() > 0)
+                {
+                  Assert(new_numbers.data() != nullptr, ExcInternalError());
+                  std::memcpy(ghosted_new_numbers.data() +
+                                n_ghosts_on_smaller_ranks,
+                              new_numbers.data(),
+                              sizeof(types::global_dof_index) *
+                                new_numbers.size());
+                }
+            }
+
+            // In case we do not carry any relevant dof (but only some remote
+            // processor), we do not need to call the renumbering. We call the
+            // version without validity check because vertex dofs will be
+            // set already in the artificial region.
+            if (relevant_dofs.n_elements() > 0)
+              Implementation::renumber_dofs(ghosted_new_numbers,
+                                            relevant_dofs,
+                                            *dof_handler,
+                                            /*check_validity=*/false);
+
+            // Since we have not updated the number cache yet, we can use the
+            // index sets contained in the DoFHandler at this stage.
+            return NumberCache(dof_handler->locally_owned_dofs_per_processor(),
+                               Utilities::MPI::this_mpi_process(
+                                 triangulation->get_communicator()));
+          }
+
+        // Now back to the more complicated case
```

Comment:
This is the case where the index sets change. Shouldn't this be in an `else` block then?

Reply:
I wanted to avoid indenting more, but you are right that this makes the code more awkward to read. I will reindent.